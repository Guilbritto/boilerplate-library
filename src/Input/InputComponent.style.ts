import styled, { css } from 'styled-components';

interface InputComponentStylesProps {
  isActive: boolean;
  variant: string;
  isEyeOn: boolean;
  error: boolean;
}

export const Container = styled.div<InputComponentStylesProps>`
  
  position: relative;
  width: auto;
  height: 48px;
  input {
    position: absolute;
    top: 0;
    left: 0;
    width: 100%;
    height: 100%;
    padding-left: .8rem;
    padding-right: 2.8rem;
    border: 2px solid #004282;
    border-radius: 4px;
    font-size: inherit;
    font-family: inherit;
    color: #2B2C32;
    outline: none;
    background: none;
    ${props => {
      if(props.isActive && props.error){
        return css`
        border-color: #E73A33;
        color: #E73A33;
        `
      }
      if(props.isActive && !props.error){
        return css`
          border-color:#2B52DD
        `
      }
      if (props.isEyeOn) {
        return css`
          -webkit-text-security: disc;
        `
      }
      if (props.isEyeOn && props.error) {
        return css`
          -webkit-text-security: disc;
          border-color: #E73A33;
        `
      }
      
      
    } };
  }
  
  label{
    position: absolute;
    left: 1rem;
    top: 35%;
    padding: 0 .5rem;
    margin-left: -.5rem;
    color: #999EA7;
    cursor: text;
    transition: top 200ms ease-in,
    left 200ms ease-in,
    font-size 200ms ease-in;
    
    ${props => props.isActive && css`
      top: -.5rem;
      font-size: .8rem;
      left: .8rem;
      background-color: #FFF;
      color: #2B52DD;
    `}
    ${props => props.error && css` 
      color: #E73A33;
    `}
  }
  svg {
    position: absolute;
    right: -2.9rem;
    top: .9rem;
    cursor: pointer;
  }
  
`;

export const AlertMessage = styled.div`
  display: flex;
  align-items: center;
  justify-content: flex-start;
  padding: 5px;
  color: #E73A33;

  svg{
    margin-right: 5px;
  }
`