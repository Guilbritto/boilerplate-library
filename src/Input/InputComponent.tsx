import React, { useEffect, useRef, useState } from 'react'

import {
    Container,
    AlertMessage,
    Input,
    Label,
    Svg
} from './InputComponent.styles'
import { InputProps } from './InputComponent.types'

import { Eye } from '@svg/Eye'
import { EyeOff } from '@svg/EyeOff'
import { AlertCircle } from '@svg/AlertCircle'
import { DefaultTheme, withTheme } from 'styled-components'

const InputComponent = ({
    value,
    variant = 'text',
    label,
    error,
    type = 'text',
    theme,
    customOnChange,
    ...rest
}: InputProps) => {
    const [isActive, setIsActive] = useState(false)
    const [inputText, setInputText] = useState(value || '')
    const [isEyeOn, setIsEyeOn] = useState(true)
    const inputRef = useRef<HTMLInputElement>(null)

    useEffect(() => {
        if (inputText) {
            setIsActive(true)
        }
    })

    useEffect(() => {
        if (!!error) {
            setIsActive(true)
        }
    }, [error, inputRef])

    const heandleFocus = () => {
        setIsActive(true)
    }

    const handleBlur = () => {
        if (inputText.length === 0) {
            setIsActive(false)
        }
    }

    const handleChange = (e: React.ChangeEvent<HTMLInputElement>) => {
        setInputText(e.target.value)
        if (customOnChange) {
            customOnChange(e)
        }
    }
    const handleEyeClick = () => {
        setIsEyeOn(!isEyeOn)
        if (isEyeOn) {
            inputRef.current?.setAttribute('type', 'text')
        } else {
            inputRef.current?.setAttribute('type', 'password')
        }
    }

    return (
        <>
            <Container
                variant={variant}
                onClick={heandleFocus}
                onBlur={handleBlur}
                data-testid="input-container"
            >
                <Input
                    {...rest}
                    isActive={isActive}
                    isEyeOn={isEyeOn && variant === 'password'}
                    error={!!error}
                >
                    <input
                        value={inputText}
                        data-testid="input"
                        onChange={evt => handleChange(evt)}
                        ref={inputRef}
                        type={variant === 'password' ? 'password' : 'text'}
                    />
                    {variant === 'password' && (
                        <Svg
                            onClick={() => handleEyeClick()}
                            data-testid="input-eye"
                        >
                            {isEyeOn ? Eye : EyeOff}
                        </Svg>
                    )}
                </Input>
                <Label
                    isActive={isActive}
                    error={!!error}
                    data-testid="input-label"
                >
                    {label}
                </Label>
            </Container>
            {error && (
                <AlertMessage data-testid="input-errorMessage">
                    {AlertCircle}
                    {error?.message}
                </AlertMessage>
            )}
        </>
    )
}

export default withTheme(InputComponent)
