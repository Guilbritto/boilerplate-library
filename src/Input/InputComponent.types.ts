import { DefaultTheme } from 'styled-components'

export interface InputProps
    extends React.InputHTMLAttributes<HTMLInputElement> {
    label: string
    value?: string
    error?: InputError
    variant?: 'password' | 'text'
    customOnChange?: (event: React.ChangeEvent<HTMLInputElement>) => void
    theme?: DefaultTheme
}

export type InputError = {
    message?: string
    icon?: string
}
