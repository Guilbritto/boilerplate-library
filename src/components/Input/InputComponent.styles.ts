import styled, { css } from 'styled-components'

interface InputComponentStylesProps {
    isActive?: boolean
    variant?: string
    isEyeOn?: boolean
    error?: boolean
}

export const AlertMessage = styled.div`
    display: flex;
    align-items: center;
    justify-content: flex-start;
    padding: 5px;
    color: ${props => props.theme.colors.suport.error};
    svg {
        margin-right: 5px;
    }
`

export const Container = styled.div<InputComponentStylesProps>`
    box-sizing: border-box;
    position: relative;
    width: auto;
    height: 48px;
`

export const Svg = styled.div`
    display: flex;
    svg {
        right: -2.9rem;
        top: 0.9rem;
        cursor: pointer;
    }
`

export const Label = styled.label<InputComponentStylesProps>`
    position: absolute;
    left: 1rem;
    top: 35%;
    padding: 0 0.5rem;
    margin-left: -0.5rem;
    color: ${props => props.theme.colors.neutral.soft};
    cursor: text;
    transition: top 200ms ease-in, left 200ms ease-in, font-size 200ms ease-in;

    ${props =>
        props.isActive &&
        css`
            top: -0.5rem;
            font-size: 0.8rem;
            left: 0.8rem;
            background-color: ${props => props.theme.colors.neutral.white};
            color: ${props => props.theme.colors.primary};
        `}
    ${props =>
        props.error &&
        css`
            color: ${props => props.theme.colors.suport.error};
        `}
`

export const Input = styled.div<InputComponentStylesProps>`
    display: flex;
    top: 0;
    left: 0;
    width: 100%;
    height: 100%;
    padding-left: 0.8rem;
    padding-right: 0.8rem;
    border: 2px solid ${props => props.theme.colors.secondary};
    border-radius: 4px;
    color: ${props => props.theme.colors.neutral.dark};
    align-items: center;
    justify-content: center;
    input {
        font-size: inherit;
        font-family: inherit;
        width: 100%;
        height: 88%;
        outline: none;
        background: none;
        border-style: none;
        &::-ms-reveal,
        &::-ms-clear {
            display: none;
        }
        ${props =>
            props.isActive &&
            props.error &&
            css`
                color: ${props => props.theme.colors.suport.error};
            `}
    }
    ${props =>
        props.error &&
        css`
            border-color: ${props => props.theme.colors.suport.error};
        `}
    ${props =>
        props.isActive &&
        props.error &&
        css`
            border-color: ${props => props.theme.colors.suport.error};
            color: ${props => props.theme.colors.suport.error};
        `}
    ${props =>
        props.isActive &&
        !props.error &&
        css`
            border-color: ${props => props.theme.colors.primary};
        `}
`
