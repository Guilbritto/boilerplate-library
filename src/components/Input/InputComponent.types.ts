import { DefaultTheme } from "styled-components";

export interface InputProps extends React.InputHTMLAttributes<HTMLInputElement>{
    label: string;
    value?: string;
    error?: InputError
    variant?: "password" | "text";
    customOnChange?: (event: React.ChangeEvent<HTMLInputElement>) => void;
}

export type InputError ={
    message?: string;
    icon?: string;
}


