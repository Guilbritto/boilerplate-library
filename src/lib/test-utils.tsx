import { render, RenderOptions } from '@testing-library/react'
import React, { ReactElement, FC } from 'react'
import { ThemeProvider } from 'styled-components'
import { defaultTheme } from '../styles/defaultTheme'

const Providers: FC = ({ children }) => {
    return <ThemeProvider theme={defaultTheme}>{children}</ThemeProvider>
}

/* eslint-disable @typescript-eslint/explicit-module-boundary-types */
const customRender = (
    ui: ReactElement,
    options?: Omit<RenderOptions, 'queries'>
) => render(ui, { wrapper: Providers, ...options })

// re-export everything
export * from '@testing-library/react'

// override render method
export { customRender as render }
