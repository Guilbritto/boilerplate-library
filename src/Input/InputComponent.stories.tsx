import React from "react";
import InputComponent from "./InputComponent";

export default {
  title: "InputComponent",
};

export const Active = () => {
  return <InputComponent label="UsuÃ¡rio" value="Text" />;
};

export const Text = () => {
  return <InputComponent label="UsuÃ¡rio" />;
};

export const Password = () => {
  return <InputComponent label="UsuÃ¡rio" variant="password" />;
};
