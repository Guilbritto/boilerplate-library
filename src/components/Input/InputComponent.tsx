import React, { useCallback, useEffect, useRef, useState } from 'react'
import {
    Container,
    AlertMessage,
    Input,
    Label,
    Svg,
    Mediumlabel
} from './InputComponent.styles'
import { InputProps } from './InputComponent.types'
import { Eye } from '@svg/Eye'
import { EyeOff } from '@svg/EyeOff'
import { AlertCircle } from '@svg/AlertCircle'
import { useTheme } from '../../hooks/useTheme'

const InputComponent = ({
    variant = 'text',
    label,
    error,
    type = 'text',
    marginTop,
    marginBottom,
    marginLeft,
    marginRight,
    width,
    inputSize = 'large',
    disabled = false,
    mask,
    value,
    ...rest
}: InputProps) => {
    const [isActive, setIsActive] = useState(false)
    const [isEyeOn, setIsEyeOn] = useState(true)
    const inputRef = useRef<HTMLInputElement>(null)
    const [filled, setFilled] = useState(false)

    const theme = useTheme()

    useEffect(() => {
        if (
            inputSize === 'large' &&
            inputRef.current?.value &&
            inputRef.current.value != ''
        ) {
            setIsActive(true)
        }
    })

    useEffect(() => {
        if (!!error && error.message) {
            setIsActive(true)
        }
    }, [error, inputRef])

    const heandleFocus = () => {
        setIsActive(true)
        inputRef.current?.focus()
    }

    const handleBlur = () => {
        if (inputSize === 'medium' || inputRef.current?.value.length === 0) {
            setIsActive(false)
        }
        if (inputRef.current?.value.length > 0) {
            setFilled(true)
        } else {
            setFilled(false)
        }
    }

    const handleEyeClick = () => {
        setIsEyeOn(!isEyeOn)
        if (isEyeOn) {
            inputRef.current?.setAttribute('type', 'text')
        } else {
            inputRef.current?.setAttribute('type', 'password')
        }
    }

    const dateMask = (
        value?: string | ReadonlyArray<string> | number | undefined
    ) => {
        if (!value) return value

        let NewValue = `${value}`

        NewValue = NewValue.replace(/\D/g, '')

        if (NewValue.length <= 3) NewValue = NewValue.replace(/(\d{2})/, '$1/')

        if (NewValue.length === 4)
            NewValue = NewValue.replace(/(\d{2})(\d{2})/, '$1/$2/')

        if (NewValue.length > 4)
            NewValue = NewValue.replace(/(\d{2})(\d{2})(\d)/, '$1/$2/$3')

        return NewValue.substring(0, 10)
    }

    const timeMask = (
        value?: string | ReadonlyArray<string> | number | undefined
    ) => {
        if (!value) return value

        let NewValue = `${value}`

        NewValue = NewValue.replace(/\D/g, '')

        if (NewValue.length <= 3) NewValue = NewValue.replace(/(\d{2})/, '$1:')

        if (NewValue.length === 4)
            NewValue = NewValue.replace(/(\d{2})(\d{2})/, '$1:$2:')

        if (NewValue.length > 4)
            NewValue = NewValue.replace(/(\d{2})(\d{2})(\d)/, '$1:$2:$3')

        return NewValue.substring(0, 8)
    }

    const handleMask = useCallback(
        (value: any) => {
            if (mask === 'date') {
                return dateMask(value)
            }

            if (mask === 'time') {
                return timeMask(value)
            }

            return value
        },
        [mask]
    )

    return (
        <div
            style={{
                marginTop,
                marginBottom,
                marginLeft,
                marginRight,
                width,
                height: inputSize === 'medium' && !!error ? '74px' : 'auto',
                display: 'flex',
                flexDirection: 'column',
                justifyContent: 'space-between'
            }}
        >
            <Container
                variant={variant}
                onClick={heandleFocus}
                onBlur={handleBlur}
                data-testid="input-container"
                theme={theme}
                disabled={disabled}
                inputSize={inputSize}
            >
                {inputSize === 'medium' && (
                    <Mediumlabel
                        isActive={isActive}
                        error={!!(error && error.message)}
                        data-testid="input-label"
                        theme={theme}
                        disabled={disabled}
                    >
                        {label}
                    </Mediumlabel>
                )}
                <Input
                    isFilled={filled}
                    isActive={isActive}
                    isEyeOn={isEyeOn && variant === 'password'}
                    error={!!(error && error.message)}
                    inputSize={inputSize}
                    theme={theme}
                    disabled={disabled}
                >
                    <input
                        data-testid="input"
                        ref={inputRef}
                        disabled={disabled}
                        type={variant === 'password' ? 'password' : 'text'}
                        value={handleMask(value)}
                        {...rest}
                    />
                    {variant === 'password' && (
                        <Svg
                            theme={theme}
                            onClick={() => handleEyeClick()}
                            data-testid="input-eye"
                        >
                            {isEyeOn ? Eye : EyeOff}
                        </Svg>
                    )}
                </Input>
                {inputSize === 'large' && (
                    <Label
                        theme={theme}
                        isActive={isActive}
                        error={!!(error && error.message)}
                        data-testid="input-label"
                        disabled={disabled}
                    >
                        {label}
                    </Label>
                )}
            </Container>
            {error && error.message && (
                <AlertMessage theme={theme} data-testid="input-errorMessage">
                    {AlertCircle}
                    {error?.message}
                </AlertMessage>
            )}
        </div>
    )
}

export default InputComponent
