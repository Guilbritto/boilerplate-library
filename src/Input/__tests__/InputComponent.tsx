import React from 'react'
import 'jest-styled-components'
import Input from '../InputComponent'
import '@testing-library/jest-dom/extend-expect'
import { fireEvent, render } from '../../lib/test-utils'
jest.mock('@svg/AlertCircle', () => {
    return () => <></>
})

describe('Input Component', () => {
    it('Should be able to match a snapshot with false parameter', () => {
        const { container } = render(
            <Input
                label="Test"
                error={{
                    message: 'Test'
                }}
                variant="text"
            />
        )

        expect(container.firstChild).toMatchSnapshot()
    })
    it('Should be able to match a snapshot with true parameter', () => {
        const { container } = render(
            <Input
                label="Test"
                error={{
                    message: 'Test'
                }}
                variant="password"
            />
        )

        expect(container.firstChild).toMatchSnapshot()
    })
    it('Should be able to check the error message', () => {
        const errorMessage = 'Error Message'
        const { getByTestId } = render(
            <Input
                error={{
                    message: errorMessage
                }}
                label="Label"
                variant="password"
            />
        )
        const alertMessage = getByTestId('input-errorMessage')
        expect(alertMessage.innerHTML).toBe(errorMessage)
    })
    it('Should be not able to check the error message when message is diferent', () => {
        const errorMessage = 'Error Message'
        const { getByTestId } = render(
            <Input
                error={{
                    message: errorMessage
                }}
                label="Label"
                variant="password"
            />
        )
        const alertMessage = getByTestId('input-errorMessage')
        expect(alertMessage.innerHTML === 'Mensagem Diferente').toBeFalsy()
    })
    it('Should be able to change a value of input', () => {
        const { getByTestId } = render(
            <Input
                error={{
                    message: ''
                }}
                label="Label"
                variant="password"
                customOnChange={jest.fn}
            />
        )
        const input = getByTestId('input')
        const value = 'value'
        fireEvent.change(input, { target: { value } })
        expect((input as HTMLInputElement).value).toBe(value)
    })
    it('Should be able to click on eye', () => {
        const { getByTestId } = render(
            <Input
                error={{
                    message: ''
                }}
                label="Label"
                variant="password"
                customOnChange={jest.fn}
            />
        )
        const eye = getByTestId('input-eye')
        const input = getByTestId('input')

        fireEvent.click(eye)
        expect((input as HTMLInputElement).type).toBe('text')
        fireEvent.click(eye)
        expect((input as HTMLInputElement).type).toBe('password')
    })
    it('Should be able to click outside the input and trigger blur', () => {
        const { getByTestId } = render(
            <Input
                error={{
                    message: ''
                }}
                label="Label"
                variant="password"
            />
        )
        const input = getByTestId('input')
        input.focus()
        expect(input).toHaveFocus()
        input.blur()
        const label = getByTestId('input-label')
        expect(label).toMatchSnapshot()
    })
})
