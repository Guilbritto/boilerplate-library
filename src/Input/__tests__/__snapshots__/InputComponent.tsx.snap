// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input Component Should be able to match a snapshot with false parameter 1`] = `
.c0 {
  position: relative;
  width: auto;
  height: 48px;
}

<div
  class="c0"
/>
`;

exports[`Input Component Should be able to match a snapshot with true parameter 1`] = `
.c0 {
  position: relative;
  width: auto;
  height: 48px;
}

<div
  class="c0"
/>
`;
